### PR TITLE
Add archive type keyword

### PR DESCRIPTION
**Pull request by Korusuke**
This addresses Issue #3 

Since upt is setting archive_type to targz for all sdist, so instead of using that extracting from URL seems better.
@Steap Should this feature be directly implemented in upt itself or is this fine?

Once this is merged I will push the changes for #4

### CODE REVIEW

**Comment by reneeotten** [comment]
> Since upt is setting archive_type to targz for all sdist, so instead of using that extracting from URL seems better. @Steap Should this feature be directly implemented in upt itself or is this fine?

That's what appeared to happen from a quick look at the ```upt``` code. I would think the ```archive_type``` could/should be set correctly in ```upt```, but let's see what @Steap thinks about this. Otherwise getting the information from the URL and doing this in the MacPorts backend would work as well.

Why the change in line 19 to ```{% endblock -%}``` ; what does the ```-``` sign do?

**Comment by Korusuke** [comment]
The `-` sign is just for deleting newline after that block as if it's not there there would be an extra line after `distname` in python ports and two blank lines after `supported_archs` in perl ports.

**Comment by reneeotten** [comment]
> The `-` sign is just for deleting newline after that block as if it's not there there would be an extra line after `distname` in python ports and two blank lines after `supported_archs` in perl ports.

Aha, okay... so that's how to avoid empty lines, thanks for the explanation ;) Let's see if the idea is to get the ```archive_type``` set in ```upt``` or if we should do this in the backend.

**Comment by Steap** [comment]
Hey! Thanks for the PR.

So, in upt, the archive_type is set to SOURCE_TARGZ even when upstream uses a ZIP file, for instance. I'll probably rename it to "SOURCE" so that it is less confusing. I could encode the filetype, but I think backends would still need to have a long "switch/case" to handle all possible cases, so that may not help that much.

Your patch looks good, but it could crash in some cases. upt_pkg.archives might be an empty list, so you'd get an IndexError when accessing the first element. You should probably move this code to upt_macports.py and have lighter code in the template. This would also make it easier to write tests. You may also want to use "upt_pkg.get_archive()" which will return the source archive, rather than archives[0].

**Comment by Korusuke** [comment]
Initially, I had written this code in `upt_macports.py` and was passing just the compression type such as `zip`, `xz`, etc. and then in the template was just appending the type to `use` and writing `yes` there.
But the problem with that was
If the type is zip we need:
<pre>
usezip          yes
</pre>
and if xz then:
<pre>
usexz           yes
</pre>
but instead what i was getting was:
<pre>
usexz            yes
</pre>
as the spaces between usezip, usexz and yes is different that was causing issues, I will look into if there's any solution this.
If I don't find anything then the py script itself can return the whole thing like `usexz           yes` 


Also about the crash due to archive[0], It will cause the same issue in checksums, right?
`pkg.upt_pkg.archives[0].sha256`

**Comment by reneeotten** [comment]
@Korusuke you can the [```.format()``` method](https://docs.python.org/3/library/string.html#format-string-syntax) and specify how to align things, so that the ```yes``` is always aligned correctly.

**Comment by reneeotten** [comment]
@Korusuke from a quick check it appears to be working as intended. However, I would encourage you to take another look at the ```archive_type``` method you added and see if that could be written in a more compact/Pythonic way. 

To get you started, I would think about it like this:
- let's get the extension of the file (i.e., the part after the last ```.```)
- then there are three scenarios: (i) the extension is in a group/list that can be used directly in the ```use_extension```, (ii) it's ```bz2``` - in that case it should be ```bzip2``` instead of ```bz2```, (iii) it's not part of the pre-defined MacPorts archive types, then return "unknown" or something like that.

In the template there are then two scenarios: (i) the archive_type/extension is unknown, (ii) it is one of the pre-defined ones add ```use_extension``` to the template unless it's ```.gz```.

File: upt_macports/tests/test_macports_package.py
Comment (Steap) [inline upt_macports/tests/test_macports_package.py:39]: Add a test with:

>  self.package.upt_pkg.archives = []

See what happens.

File: upt_macports/upt_macports.py
Comment (Steap) [inline upt_macports/upt_macports.py:73]: This is fine, but maybe we could do something like;:

```
mappings = {
    '.tar.gz': 'targz',
    '.7z': '7z',
    ...
}
...
for extension, fmt in mappings.items():
    if archive_url.endswith(extension):
        return fmt
return 'undefined'
```

What do you think? There might be better variable names :)

**Review by Steap — CHANGES_REQUESTED** [review]

**Comment by Steap** [comment]
> * let's get the extension of the file (i.e., the part after the last `.`)

In the case of "foo-13.37.tar.bz2", the extension is actually "tar.bz2", so it is not always the part after the last '.', or am I misunderstanding something?

File: upt_macports/templates/base.Portfile
Comment (Steap) [inline upt_macports/templates/base.Portfile]: This empty if is weird, maybe you could...

File: upt_macports/templates/base.Portfile
Comment (Steap) [inline upt_macports/templates/base.Portfile:24]: ... use "{{ if pkg.archive_type != 'targz' }}" here. What do you think?

**Comment by reneeotten** [comment]
> > * let's get the extension of the file (i.e., the part after the last `.`)
> 
> In the case of "foo-13.37.tar.bz2", the extension is actually "tar.bz2", so it is not always the part after the last '.', or am I misunderstanding something?

You're certainly right about that, but please take a look in the [MacPorts guide](https://guide.macports.org/#reference.phases.extract) and how the ```use_???``` are defined. You'll see that ```use_bzip2``` considers ```tar.bz2``` as the ```extract.suffix``` and ```use_xz``` corresponds to ```tar.xz```. So for all practical purposes I think that the thing we care about here is "the part after the last ```.```); of course we don't have to call it the file "extension" ;)

File: upt_macports/upt_macports.py
Comment (Korusuke) [reply upt_macports/upt_macports.py:73]: If using just the last `.` is fine then maybe we can do just this :

```
def archive_type(self):
        archive_keyword = {
            'gz': 'targz',
            '7z': '7z',
            'bz2': 'bzip2',
            'lzma': 'lzma',
            'tar': 'tar',
            'zip': 'zip',
            'xz': 'xz'
        }
        archive_url = self.upt_pkg.get_archive().url
        archive_type = archive_url.split('.')[-1]

        return archive_keyword.get(archive_type, 'unknown')
```

File: upt_macports/tests/test_macports_package.py
Comment (Korusuke) [reply upt_macports/tests/test_macports_package.py:39]: It raises an ArchiveUnavailable error as expected, I have added a test for the same.

```
def test_no_archive(self):
        self.package.upt_pkg.archives = []
        with self.assertRaises(upt.ArchiveUnavailable):
            self.package.archive_type
```

Comment (Steap) [reply upt_macports/tests/test_macports_package.py:39]: But should "self.package.archive_type" raise an exception? Shouldn't it rather handle it and return 'unknown'?

File: upt_macports/upt_macports.py
Comment (Steap) [reply upt_macports/upt_macports.py:73]: Yeah, obviously using dict.get() is better :)

**Comment by Steap** [comment]
> > > * let's get the extension of the file (i.e., the part after the last `.`)
> > 
> > 
> > In the case of "foo-13.37.tar.bz2", the extension is actually "tar.bz2", so it is not always the part after the last '.', or am I misunderstanding something?
> 
> You're certainly right about that, but please take a look in the [MacPorts guide](https://guide.macports.org/#reference.phases.extract) and how the `use_???` are defined. You'll see that `use_bzip2` considers `tar.bz2` as the `extract.suffix` and `use_xz` corresponds to `tar.xz`. So for all practical purposes I think that the thing we care about here is "the part after the last `.`); of course we don't have to call it the file "extension" ;)

OK, makes sense.

File: upt_macports/tests/test_macports_package.py
Comment (Korusuke) [reply upt_macports/tests/test_macports_package.py:39]: If we handle the exception and generate the portfile then many details will be missing as checksums will be missing/wrong too.
So should we really generate the portfile when the main thing is missing?
I agree that at least a few details will be there and that can be used as an outline by the maintainer.

Comment (Steap) [reply upt_macports/tests/test_macports_package.py:39]: The checksums would be missing, not wrong.

We should probably "do our best". The goal of upt is to create roughly 90% of the package. If for some reason we can only create 50%, well, the maintainer still has something to work with, and saved a few minutes of their life.

We should probably log this as an error, though.

File: upt_macports/upt_macports.py
Comment (reneeotten) [reply upt_macports/upt_macports.py:73]: Agreed! For the sake of consistency I would just use ```'gz': 'gz'``` as this is no different than for example the situation for ```xz```.

File: upt_macports/tests/test_macports_package.py
Comment (reneeotten) [reply upt_macports/tests/test_macports_package.py:39]: > We should probably log this as an error, though.
Yes, that would be good indeed (there are probably other instances to log as error/warning; for example when the license cannot be translated to MacPorts). Should that be part of this PR or should we look at the logging by ```upt_macports``` separately?

Comment (Korusuke) [reply upt_macports/tests/test_macports_package.py:39]: Ok so if the archive is not found then at this point should we log 'Archive not found on upstream' or should we say 'Archive detection failed as no archive found' as the latter is appropriate as per the place where it is being logged but even the former is not wrong.

And also as of now, I am logging it like this 
```
try:
            archive_url = self.upt_pkg.get_archive().url
            archive_type = archive_url.split('.')[-1]
            return archive_keyword.get(archive_type, 'unknown')
except upt.ArchiveUnavailable:
            self.logger.error(f'Archive detection failed as no archive found')
            return 'unknown'
```
and the test has been changed accordingly.
```
self.package.upt_pkg.archives = []
expected = 'unknown'
self.assertEqual(self.package.archive_type, expected)
```

Is it fine now? :)

Also as of the other logging goes I guess from now on for every feature wherever required we will add the logging and for the existing code, we can add the logging in another PR.

Comment (reneeotten) [reply upt_macports/tests/test_macports_package.py:39]: hi @Korusuke, ehm I don't really see the change you mention above in the PR... but perhaps you were waiting for a response before adding it?

I would like to avoid two ```return``` statements, so what about:
```
try:
    archive_url = self.upt_pkg.get_archive().url
    archive_type = archive_url.split('.')[-1]
except upt.ArchiveUnavailable:
    self.logger.error(f'Detection of archive type failed: no archive found')
    archive_type = None
finally:
    return archive_keyword.get(archive_type, 'unknown')
```

Additionally, as I commented earlier I would not treat ```gz``` differently in the ```archive_keyword``` dictionary than ```xz```; just return ```gz``` there as well.

**Review by reneeotten — CHANGES_REQUESTED** [review]
This should be the last round of comments - I'd suggest to wait for @Steap his reaction so everything can incorporated at once and let's get this merged!

@Korusuke thanks for the quick turnaround of changes and I'm sorry for suggesting more changes... There are always multiple ways of coding things and even though it takes a little longer before merging, it never hurts to discuss a bit and finish up with the "best" solution!

File: upt_macports/tests/test_macports_package.py
Comment (Steap) [reply upt_macports/tests/test_macports_package.py:39]: So we could use "self.upt_pkg.get_archive().filename" rather than "self.upt_pkg.get_archive().url", but both work :)

I like the simple try/except better than the try/except/finally with a single return statement, but I'm ok with both as long as we have good tests.

File: upt_macports/upt_macports.py
Comment (Steap) [reply upt_macports/upt_macports.py:73]: I'm confused, don't we have a special case for "tar.gz" files?

**Comment by Steap** [comment]
@Korusuke I agree with most of Renee's comments, so feel free to update the PR.

**Comment by Steap** [comment]
Oh btw, can't we add "Fixes: #3" in the commit log so that the bug is closed automatically once this is merged?

File: upt_macports/tests/test_macports_package.py
Comment (reneeotten) [reply upt_macports/tests/test_macports_package.py:39]: fine with me, leave the ```try....except``` as is, but please do change the logging message. Using ```self.upt_pkg.get_archive().filename``` would indeed make sense.

File: upt_macports/upt_macports.py
Comment (reneeotten) [reply upt_macports/upt_macports.py:73]: Yes, we do.

What I mean is that what is returned shouldn't be ```targz``` in my opinion (if we follow that logic ```xz``` should return ```tarxz```. So what I am proposing is -for consistency - to change ```'gz': 'targz',``` to ```'gz': 'gz',``` and in the template use ```{% elif pkg.archive_type != 'gz' %}```

File: upt_macports/tests/test_macports_package.py
Comment (Korusuke) [reply upt_macports/tests/test_macports_package.py:39]: Yeah using `self.upt_pkg.get_archive().filename` make more sense, I had missed that it exists as my local copy of upt was old and did not have the filename, filepath properties :)

File: upt_macports/upt_macports.py
Comment (reneeotten) [inline upt_macports/upt_macports.py]: I am not a fan of this indentation, but I know you did that to get the text continued correctly. I would not be against just keeping this on one line and violating the "80-length" rule a bit, but perhaps the tests would complain about that....  Another option would be:
```
self.logger.error(f'Detection of archive type failed: no archive ' \
                   'found')
```

besides that (even with this) I would be happy to merge!

**Comment by mojca** [comment]
In the test you can explicitly say that it's ok to violate the line length.

File: upt_macports/upt_macports.py
Comment (Korusuke) [reply upt_macports/upt_macports.py]: Ok so I will keep it normal (violate 80-length) as that looks better and then maybe in other PR or if you are fine then in this same PR change the test to allow that rule violation as @mojca said.

File: upt_macports/upt_macports.py
Comment (Steap) [inline upt_macports/upt_macports.py]: Do not use an f-string when it is not needed. I also think this should rather read something like "Could not determine the type of the source archive". Writing "no archive found" makes me think upt-macports was looking for the archive, while it was upt-pypi which failed to find one.

Also, if the line is too long, you can add "# noqa" at the end to disable the linter for this particular line.

Comment (Korusuke) [reply upt_macports/upt_macports.py]: > Also, if the line is too long, you can add "# noqa" at the end to disable the linter for this particular line.

Oo I had no idea about this, I thought global test settings bust be changed.

I have done the new changes as suggested.

**Review by reneeotten — APPROVED** [review]

**Comment by reneeotten** [comment]
Great - a concise function to do the work, tests added to make sure everything works and an updated template! What else can one request - in it goes ;)

**Pull request merged**
